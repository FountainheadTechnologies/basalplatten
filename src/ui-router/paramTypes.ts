import { ParamTypeDefinition } from '@uirouter/core';

import {
  is,
  equals,
  pipe,
  pickBy,
  isNil,
  toPairs,
  fromPairs,
  map,
  join,
  split
} from 'ramda';

export const where: ParamTypeDefinition = {
  raw: true,

  equals,

  is: is(Object),

  pattern: /\w+~?:\w+\|*|^$/,

  encode: pipe(
    pickBy(value => {
      if (value instanceof Array) {
        return value.length > 0;
      }

      return !isNil(value);
    }),

    toPairs,

    map<[string, string | string[]], string>(([key, value]) => {
      if (value instanceof Array) {
        key = `~${key}`;
        value = value.map(encodeURIComponent);
      } else {
        value = encodeURIComponent(value);
      }

      return [key, value].join(':');
    }),

    join('!'),
  ),

  decode: (pipe as any)(
    split('!'),

    map((pair: string) => {
      var [key, value] = pair.split(':');

      if (key[0] === '~') {
        return [
          key.slice(1, key.length),
          value.split(',').map(decodeURIComponent)
        ];
      }

      return [key, decodeURIComponent(value)];
    }),

    fromPairs,
  )
};

export const order: ParamTypeDefinition = {
  raw: true,

  equals,

  is: is(Object),

  encode: (value) => {
    var keys = Object.keys(value);

    if (!keys.length) {
      return '';
    }

    return keys
      .map(key => [key, value[key]].join(':'))
      .join(',');
  },

  decode: (value) => {
    var [field, direction] = value.split(':');
    return { [field]: direction };
  },

  pattern: /\w+:(?:asc|desc)|^$/
}
