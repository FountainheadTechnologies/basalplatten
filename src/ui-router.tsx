import React from 'react';
import { UIRouterReact, servicesPlugin, pushStateLocationPlugin } from '@uirouter/react';
import { UIRouterRx } from '@uirouter/rx';
import * as paramTypes from './ui-router/paramTypes';

import notification from 'antd/lib/notification';
import 'antd/lib/notification/style';

export const buildRouter = (): UIRouterReact => {
  var router = new UIRouterReact();

  router.plugin(servicesPlugin);
  router.plugin(pushStateLocationPlugin);
  router.plugin(UIRouterRx);

  router.stateService.defaultErrorHandler(err => {
    // 'Transition Superseded' errors should not be handled
    if (err.type && err.type === 2) {
      return;
    }

    notification.error({
      message: 'State Transition Error',

      description: (
        <div>
          <p>The following error was encountered during a UI state transition:</p>
          <code>{err.stack}</code>
        </div>
      )
    });

    throw err;
  });

  router.urlRouter.otherwise((matchValue, url, router) => {
    const description = url ?
      `The URL ${url.path} does not exist. You have been redirected.` :
      `You attempted to access an unknown URL. You have been redirected.`;

    notification.info({
      message: 'Unknown URL',
      description
    });

    return '/';
  });

  Object.keys(paramTypes).forEach(name => {
    router.urlMatcherFactory.type(name, (paramTypes as any)[name]);
  });

  return router;
}

