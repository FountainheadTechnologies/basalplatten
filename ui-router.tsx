import * as React from 'react';
import { UIRouterReact, servicesPlugin, pushStateLocationPlugin } from '@uirouter/react';
import { UIRouterRx } from '@uirouter/rx';
import notification from 'antd/lib/notification';
import * as paramTypes from './ui-router/paramTypes';
import 'antd/lib/notification/style';

export const buildRouter = (): UIRouterReact => {
  var router = new UIRouterReact();

  router.plugin(servicesPlugin);
  router.plugin(pushStateLocationPlugin);
  router.plugin(UIRouterRx);

  router.stateService.defaultErrorHandler(err => {
    notification.error({
      message: 'State Transition Error',

      description: (
        <div>
          <p>The following error was encountered during a UI state transition:</p>
          <code>{err.stack}</code>
        </div>
      )
    });

    throw err;
  });

  router.urlRouter.otherwise((matchValue, url, router) => {
    const description = url ?
      `The URL ${url.path} does not exist. You have been redirected.` :
      `You attempted to access an unknown URL. You have been redirected.`;

    notification.info({
      message: 'Unknown URL',
      description
    });

    return '/';
  });

  Object.keys(paramTypes).forEach(name => {
    router.urlMatcherFactory.type(name, (paramTypes as any)[name]);
  });

  return router;
}

