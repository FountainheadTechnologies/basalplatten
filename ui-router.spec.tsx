import {buildRouter} from './ui-router';
import notification from 'antd/lib/notification';
import * as paramTypes from './ui-router/paramTypes';

it('adds a default error handler', async () => {
  var router = buildRouter();

  router.stateRegistry.register({
    name: 'badState',
    resolve: {
      willThrow: () => {
        throw 'whoops';
      }
    }
  });

  try {
    await router.stateService.go('badState');
  } catch (e) {}

  expect(notification.error).toHaveBeenCalled();
  var options = (notification.error as jest.Mock<any>).mock.calls[0][0];

  expect(options.message).toBe('State Transition Error');
  expect(options.description).toMatchSnapshot();
});

it('adds a URL redirection handler', () => {
  var router = buildRouter();

  try {
    window.location.pathname = '/test';
    router.urlRouter.sync();
  } catch (e) {}

  expect(notification.info).toHaveBeenCalled();
  var options = (notification.info as jest.Mock<any>).mock.calls[0][0];

  expect(options.message).toBe('Unknown URL');
  expect(options.description).toMatchSnapshot();
});

it('registers custom parameter type definitions', () => {
  var router = buildRouter();

  var registeredTypes = Object.keys(router.urlMatcherFactory.paramTypes.types);
  var definedTypes = Object.keys(paramTypes);

  expect(registeredTypes).toEqual(definedTypes);
});
