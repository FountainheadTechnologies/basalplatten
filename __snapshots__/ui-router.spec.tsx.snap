// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a URL redirection handler 1`] = `"The URL blank does not exist. You have been redirected to the home page."`;

exports[`adds a default error handler 1`] = `
<div>
  <p>
    The following error was encountered during a UI state transition:
  </p>
  <code />
</div>
`;
